### ISSUE TITLE
Experiment: Tri-Loss Topological Routing (Intra-Modal Regularization via Cross-Attention Projected Features)

### ISSUE THREAD

**Issue by iremurhan** [iremurhan/tez_v2_clean#34]
# Experiment: Tri-Loss Topological Routing — Intra-Modal Regularization via Cross-Attention Projected Features

**Issue #34** · `exp/i34-dual-loss-routing`

**Parent:** Issue #30 (Text-Guided Visual Pooling via Cross-Attention)

---

## Motivation

Issue #30 routed the Cross-Attention pooled output $F_{\text{image}}$ directly into the cross-modal InfoNCE loss. This collapsed the representational geometry of the shared embedding space by replacing the ViT's native patch-aggregated representation with a text-conditioned one. Result: **Test i2t R@1 degraded from 70.62 → 64.44** (−6.18 absolute).

**Root Cause Hypothesis:** The cross-attention output is text-conditioned by construction. Feeding it into the *same* contrastive objective that aligns image↔text creates a circular dependency: the loss gradient simultaneously pushes $F_{\text{image}}$ toward the text embedding *and* pulls the text embedding toward $F_{\text{image}}$, which is already a function of the text query. This short-circuits the contrastive signal.

**Architectural Pivot:** Decouple the representation streams into three loss terms with distinct gradient pathways:

| Signal | Source Tensor | Loss Term | Gradient Target |
|---|---|---|---|
| Cross-Modal (Image↔Text) | ViT CLS token (direct) | $\mathcal{L}_{\text{cross}}$ (InfoNCE) | ViT encoder, text encoder |
| Intra-Modal Image (Image↔Image) | Cross-Attention pooled $F_{\text{image}}$ | $\mathcal{L}_{\text{i2i}}$ (InfoNCE) | Cross-Attention module, `W_patch` |
| Intra-Modal Text (Text↔Text) | Projected text CLS via `W_text` | $\mathcal{L}_{\text{t2t}}$ (InfoNCE) | `W_text` projection layer |

Total loss:

$$\mathcal{L} = \mathcal{L}_{\text{cross}} + \lambda_1 \cdot \mathcal{L}_{\text{i2i}} + \lambda_2 \cdot \mathcal{L}_{\text{t2t}}$$

with $\lambda_1 = 0.5$, $\lambda_2 = 0.5$ (tunable independently).

---

## Hyperparameters

| Parameter | Value | Notes |
|---|---|---|
| $\lambda_1$ (i2i weight) | 0.5 | Tunable |
| $\lambda_2$ (t2t weight) | 0.5 | Tunable |
| $\tau_{\text{cross}}$ | Learnable (CLIP `logit_scale`) | Inherited from CLIP, unchanged |
| $\tau_{\text{intra}}$ | 0.07 (fixed) | Shared by both $\mathcal{L}_{\text{i2i}}$ and $\mathcal{L}_{\text{t2t}}$. Fixed because intra-modal self-identity positives always have cosine similarity = 1.0 after L2-norm, leaving no distribution shift for a learnable τ to track. A free τ risks trivializing the loss by scaling all logits toward zero. |

---

### Branch A — Cross-Modal Loss (ViT Direct Path)

```
ViT Encoder
  Input:  image pixels
  Output: patch_embeddings    — B × N × 1024   (N = num_patches, e.g. 257 for ViT-L/14 with CLS)

Extract CLS token:
  patch_embeddings[:, 0, :]   — B × 1024

Project to shared dim via CLIP's existing visual projection:
  visual_proj(cls_token)      — B × D_clip     (D_clip = 768 for ViT-L/14)

L2-normalize:
  F_vit                       — B × D_clip

Text Encoder:
  text_cls                    — B × D_clip     (already L2-normalized by CLIP)

Cross-Modal InfoNCE:
  logits_cross = F_vit @ text_cls.T / τ_cross   — B × B
```

### Branch B — Intra-Modal Image Loss (Cross-Attention Pooled Path)

```
ViT patch embeddings (excluding CLS):
  patches = patch_embeddings[:, 1:, :]    — B × (N-1) × 1024

Project patches to d_shared:
  K = V = W_patch(patches)                — B × (N-1) × 768

Text CLS token (from text encoder, pre-projection):
  text_cls_raw                            — B × D_text (768)

Project text CLS to d_shared:
  text_proj = W_text(text_cls_raw)        — B × 768
  Q = text_proj.unsqueeze(1)              — B × 1 × 768

Cross-Attention:
  attn_out = softmax(Q @ K^T / √768) @ V — B × 1 × 768

Squeeze + L2-normalize:
  F_image = attn_out.squeeze(1)           — B × 768
  F_image_norm = L2_norm(F_image)         — B × 768

Intra-Modal Image InfoNCE:
  logits_i2i = F_image_norm @ F_image_norm.T / τ_intra   — B × B
  labels_i2i = torch.arange(B)
```

### Branch C — Intra-Modal Text Loss (Projected Text CLS Path)

```
Projected text CLS (same tensor as Cross-Attention query, before unsqueeze):
  F_text_proj = W_text(text_cls_raw)      — B × 768   (shared computation with Branch B)

L2-normalize:
  F_text_proj_norm                        — B × 768

Intra-Modal Text InfoNCE:
  logits_t2t = F_text_proj_norm @ F_text_proj_norm.T / τ_intra   — B × B
  labels_t2t = torch.arange(B)
```

> **Note on Intra-Modal Positives:** Each sample is its own positive (diagonal); all other batch members are negatives. For i2i this forces the cross-attention module to produce discriminative, image-specific features. For t2t this regularizes the `W_text` projection to preserve caption-level discriminability — preventing all projected text embeddings from collapsing to a common subspace.

> **Note on `W_text` gradient flow:** `W_text` receives gradients from **both** $\mathcal{L}_{\text{i2i}}$ (through the cross-attention query path Q) and $\mathcal{L}_{\text{t2t}}$ (directly through `F_text_proj`). This dual supervisory signal constrains `W_text` to produce projections that are simultaneously good attention queries (i2i) and individually discriminative (t2t).

---

## Computation Graph — Gradient Pathways

```
                         ┌─────────────┐
                         │ ViT Encoder │
                         └──────┬──────┘
                                │
                    ┌───────────┴───────────┐
                    │ B × N × 1024          │
               ┌────┴────┐           ┌──────┴──────┐
               │ CLS [:, 0]          │ Patches [:, 1:]
               │ B × 1024│           │ B×(N-1)×1024│
               └────┬────┘           └──────┬──────┘
                    │                       │
             visual_proj              W_patch (linear)
                    │                       │
               B × 768               B×(N-1)×768 ← K, V
                    │                       │
                 L2-norm                    │
                    │                       │
                 F_vit ──────┐              │
                             │              │
                             │   ┌──────────┴──────────┐
                             │   │  Cross-Attention    │
    ┌──────────────┐         │   │  Q @ K^T / √d → V   │
    │ Text Encoder │         │   └──────────┬──────────┘
    └──────┬───────┘         │              │
           │                 │         B × 1 × 768
      B × D_text             │              │
           │                 │          squeeze(1)
      ┌────┴─────┐           │              │
      │          │           │         B × 768
   L2-norm    W_text         │              │
      │       (linear)       │          L2-norm
      │          │           │              │
   F_text    text_proj       │        F_image_norm
   B × 768   B × 768         │              │
      │          │           │              │
      │       L2-norm        │         ┌────┴────┐
      │          │           │         │ L_i2i   │
      │    F_text_proj_norm  │         │ (InfoNCE│
      │       B × 768        │         │ i↔i)    │
      │          │           │         └─────────┘
      │     ┌────┴────┐      │
      │     │ L_t2t   │      │
      │     │ (InfoNCE│      │
      │     │ t↔t)    │      │
      │     └─────────┘      │
      │                      │
      └────────┬──────────── ┘
               │
          ┌────┴────┐
          │ L_cross │
          │ (InfoNCE│
          │ i↔t)    │
          └─────────┘

  L_total = L_cross + λ₁·L_i2i + λ₂·L_t2t
```

---

## Tasks

- [ ] **T1.** Branch off `exp/i30-text-guided-fusion` → `exp/i32-dual-loss-routing`.
- [ ] **T2.** Retain all existing Cross-Attention modules (`W_patch`, `W_text`, attention) from Issue #30. No architectural changes to these layers.
- [ ] **T3.** Restore the ViT CLS → `visual_proj` → L2-norm path as the primary input to $\mathcal{L}_{\text{cross}}$ (the original CLIP path).
- [ ] **T4.** Route Cross-Attention output $F_{\text{image}}$ exclusively to $\mathcal{L}_{\text{i2i}}$.
- [ ] **T5.** Route `W_text(text_cls_raw)` (the projected text CLS — same tensor used as cross-attention query Q) to $\mathcal{L}_{\text{t2t}}$.
- [ ] **T6.** Implement $\mathcal{L}_{\text{i2i}}$ and $\mathcal{L}_{\text{t2t}}$ as symmetric InfoNCE with in-batch self-identity positives and a fixed shared temperature $\tau_{\text{intra}} = 0.07$.
- [ ] **T7.** Compose total loss: $\mathcal{L} = \mathcal{L}_{\text{cross}} + \lambda_1 \cdot \mathcal{L}_{\text{i2i}} + \lambda_2 \cdot \mathcal{L}_{\text{t2t}}$, with $\lambda_1 = 0.5$, $\lambda_2 = 0.5$.
- [ ] **T8.** Log `loss_cross`, `loss_i2i`, `loss_t2t`, `loss_total` to wandb per step.
- [ ] **T9.** Validate: confirm Test i2t R@1 ≥ 70.0 (baseline recovery) and monitor intra-modal metrics.

---

## Evaluation Criteria

| Metric | Baseline (no fusion) | Issue #30 (broken) | Target (Issue #32) |
|---|---|---|---|
| Test i2t R@1 | 70.62 | 64.44 | ≥ 70.0 |
| Test t2i R@1 | — | — | ≥ baseline |
| $\mathcal{L}_{\text{i2i}}$ convergence | N/A | N/A | Monotonic decrease over first 5 epochs |
| $\mathcal{L}_{\text{t2t}}$ convergence | N/A | N/A | Monotonic decrease over first 5 epochs |
